### ISSUE TITLE
Font merging

### ISSUE THREAD

**Issue by rokups** [vurtun/nuklear#413]
How are we supposed to merge fonts? I spent quite some time looking through code trying to figure it out but seems there is a chicken and egg problem or something. This is what i have so far:

```cpp
    nk_font_atlas_init_default(&_nk.atlas);
    nk_font_atlas_begin(&_nk.atlas);
    _nk.atlas.default_font = nk_font_atlas_add_default(&_nk.atlas, default_font_size, 0);
    int w, h;
    nk_font_atlas_bake(&_nk.atlas, &w, &h, NK_FONT_ATLAS_RGBA32);  // Without calling this we will have no baked font.

    // <...>

    struct nk_font_config config = nk_font_config(size);
    config.range = ranges;
    config.ttf_data_owned_by_atlas = 0;
    config.merge_mode = 1;
    config.font = &_nk.atlas.default_font->info;  // baked fond required, otherwise this will fail.
    nk_font_atlas_add_from_memory(&_nk.atlas, &data.Front(), bytes_len, size, &config);

    int w, h;
    const void* image = nk_font_atlas_bake(&_nk.atlas, &w, &h, NK_FONT_ATLAS_RGBA32);

    _font_texture = <...>;

    nk_font_atlas_end(&_nk.atlas, nk_handle_ptr(_font_texture.Get()), &_nk.null_texture);
    if (_nk.atlas.default_font)
        nk_style_set_font(&_nk.ctx, &_nk.atlas.default_font->handle);
```

I am merging font-awesome font into default one this way but font-awesome icons still show up as question marks. What is the right way to merge fonts?

One more question: As i understand font setup should be done this way:
```c
nk_font_atlas_begin();
// add ALL fonts here
nk_font_atlas_bake();
nk_font_atlas_end();
```

Is it possible to later on extend font atlas with new fonts? Ideally without needing to recreate atlas with all the old fonts + new one?

**Comment by rokups** [comment]
@vurtun ping. You probably missed this issue. Any comment on how to do font merging?

**Comment by vurtun** [comment]
Could you try it the other way around first:

```c
    struct nk_font_config config = nk_font_config(size);
    config.range = ranges;
    config.ttf_data_owned_by_atlas = 0;
    config.merge_mode = 1;
    config.font = &_nk.atlas.default_font->info;  // baked fond required, otherwise this will fail.
    nk_font_atlas_add_from_memory(&_nk.atlas, &data.Front(), bytes_len, size, &config);
```
and then:

```c
_nk.atlas.default_font = nk_font_atlas_add_default(&_nk.atlas, default_font_size, 0);
```
If that does not work I need to dive deeper.

**Comment by rokups** [comment]
No dice. As i understand font baking has to be done twice right? This is what i tried:

1. Add default font
2. Bake fonts
3. Merge icon font
5. Re-add default font (optional)
6. Bake fonts

```cpp
#define ICON_MIN_FA 0xf000
#define ICON_MAX_FA 0xf2e0
void NuklearUI::TestFonts()
{
    // Step 1: add default font
    nk_font_atlas_init_default(&_atlas);
    nk_font_atlas_begin(&_atlas);
    _atlas.default_font = nk_font_atlas_add_default(&_atlas, 13.f, 0);

    // Step 2: bake default font
    {
        int w, h;
        const void* image = nk_font_atlas_bake(&_atlas, &w, &h, NK_FONT_ATLAS_RGBA32);

        _font_texture = context_->CreateObject<Texture2D>();
        _font_texture->SetNumLevels(1);
        _font_texture->SetSize(w, h, Graphics::GetRGBAFormat());
        _font_texture->SetData(0, 0, 0, w, h, image);

        nk_font_atlas_end(&_atlas, nk_handle_ptr(_font_texture.Get()), &_draw_null_texture);
        if (_atlas.default_font)
            nk_style_set_font(&_nk, &_atlas.default_font->handle);
    }

    // Step 3: merge icon font
    auto font_path = "fontawesome-webfont.ttf";
    auto size = 13.f;
    static nk_rune ranges[] = {ICON_MIN_FA, ICON_MAX_FA, 0};

    if (auto font_file = GetSubsystem<ResourceCache>()->GetFile(font_path))
    {
        PODVector<uint8_t> data;
        data.Resize(font_file->GetSize());
        auto bytes_len = font_file->Read(&data.Front(), data.Size());

        struct nk_font_config config = nk_font_config(size);
        config.range = ranges;
        config.ttf_data_owned_by_atlas = 0;
        config.merge_mode = 1;
        config.font = &_atlas.default_font->info;  // baked fond required, otherwise this will fail.
        _atlas.default_font = nk_font_atlas_add_from_memory(&_atlas, &data.Front(), bytes_len, size, &config);
    }

    // Step 5: re-add default font
    // Uncommenting next line makes icon font replace all the text.
    // If it is left commented-out then ranges from merged font are not recognized and question marks are displayed.
    // _atlas.default_font = nk_font_atlas_add_default(&_atlas, 13.f, 0);

    // Step 6: bake merged fonts
    {
        int w, h;
        const void* image = nk_font_atlas_bake(&_atlas, &w, &h, NK_FONT_ATLAS_RGBA32);

        _font_texture = context_->CreateObject<Texture2D>();
        _font_texture->SetNumLevels(1);
        _font_texture->SetSize(w, h, Graphics::GetRGBAFormat());
        _font_texture->SetData(0, 0, 0, w, h, image);

        nk_font_atlas_end(&_atlas, nk_handle_ptr(_font_texture.Get()), &_draw_null_texture);
        if (_atlas.default_font)
            nk_style_set_font(&_nk, &_atlas.default_font->handle);
    }
}
```

As you can see i tried to set result of `nk_font_atlas_add_from_memory(icon_font)` as default font. I expected it to return merged result of default font and icon font. It does not have any effect though - normal text works but icons are replaced with question marks.

Code i test with:
```c
    #define ICON_FA_STAR u8"\uf005"
    #define ICON_FA_CHECK u8"\uf00c"

    nk_labelf(*_nk, NK_TEXT_ALIGN_LEFT, "We can mix icons %s with %s text.", ICON_FA_STAR, ICON_FA_CHECK);
```

**Comment by vurtun** [comment]
Oh no that is wrong. Everything is one pass on the API side of things (out of my head):

```c
    
    nk_font_atlas_init_default(&atlas);
    nk_font_atlas_begin(&atlas);
    {
        /* I.) Add default font */
         _atlas.default_font = nk_font_atlas_add_default(&_atlas, 13.f, 0);
        /* II.) Add fontawesome with merge flag inside config */
        auto const size = 13.f;
        auto const font_path = "fontawesome-webfont.ttf";
        static const nk_rune ranges[] = {ICON_MIN_FA, ICON_MAX_FA, 0};
        if (auto font_file = GetSubsystem<ResourceCache>()->GetFile(font_path))
        {
            PODVector<uint8_t> data;
            data.Resize(font_file->GetSize());
            auto bytes_len = font_file->Read(&data.Front(), data.Size());

            struct nk_font_config config = nk_font_config(size);
            config.range = ranges;
            config.ttf_data_owned_by_atlas = 0;
            config.merge_mode = 1;
            config.font = &_atlas.default_font->info;  // baked fond required, otherwise this will fail.
            _atlas.default_font = nk_font_atlas_add_from_memory(&_atlas, &data.Front(), bytes_len, size, &config);
        }
        /* III.) Bake and upload font atlas texture */
        const void* img = nk_font_atlas_bake(&_atlas, &w, &h, NK_FONT_ATLAS_RGBA32);
        _font_texture = context_->CreateObject<Texture2D>();
        _font_texture->SetNumLevels(1);
        _font_texture->SetSize(w, h, Graphics::GetRGBAFormat());
        _font_texture->SetData(0, 0, 0, w, h, image);
    }
    /* IV.) End baking process by passing upload texture handle */
    nk_font_atlas_end(&_atlas, nk_handle_ptr(_font_texture.Get()), &_draw_null_texture);
````

**Comment by rokups** [comment]
Oh awesome that it is possible to bake everything just once. Whatever i did before caused crash in `nk_font_atlas_add_from_memory()` if default font was not baked. Also your comment in previous code snippet `baked font required, otherwise this will fail.` sort of assured that was a good idea.

I tested this snippet. It crashes when baking fonts at line `glyph = &glyphs[dst_font->glyph_offset + (unsigned int)glyph_count];`. Then i noticed code is missing this:

```c
    config.ttf_data_owned_by_atlas = 0;
    config.font = &_atlas.default_font->info;  // baked font required, otherwise this will fail.
```

Added these two lines and test application runs, but i still get question marks in place of icons.

Maybe i am testing it wrong? Icons are defined as utf-8 code points. This is what nuklear expects right? I tested same thing with ImGUI and with proper font setup `ImGui::Text("We can mix icons %s with %s text.", ICON_FA_STAR, ICON_FA_CHECK);` properly prints icons in the middle of text.

**Comment by fr500** [comment]
Did you figure this out?
I want to support fontawesome / emojis in my application and I'm wondering if it's gonna work or not.

**Comment by vurtun** [comment]
Thanks for the issue. I went ahead I fixed it in commit: 6a13addb21e49874c75ddcdda42b17c578f3ff9d



My demo code:

```c
    {struct nk_font_atlas *atlas;
    struct nk_font_config cfg = nk_font_config(16);
    static const nk_rune icon_range[] = {ICON_MIN_FA, ICON_MAX_FA,0};
    cfg.range = icon_range;
    cfg.merge_mode = 1;

    nk_sdl_font_stash_begin(&atlas);
    struct nk_font *roboto = nk_font_atlas_add_from_file(atlas, "../../../extra_font/Roboto-Regular.ttf", 16, 0);
    struct nk_font *awesome = nk_font_atlas_add_from_file(atlas, "../../../extra_font/fontawesome-webfont.ttf", 13, &cfg);
    nk_sdl_font_stash_end();
    /*nk_style_load_all_cursors(ctx, atlas->cursors);*/
    nk_style_set_font(ctx, &roboto->handle);}
```
Using font-awesome and [IconFontCppHeaders](https://github.com/juliettef/IconFontCppHeaders) with:

```c
if (nk_button_label(ctx, ICON_FA_FILE_O " File"))
    fprintf(stdout, "button pressed\n");
```
I get:

![merge](https://user-images.githubusercontent.com/8057201/32840911-4ad583bc-ca19-11e7-9618-9c5732470080.png)
